core/src/main/java/org/apache/iceberg/util/PartitionMap.java

```diff
@@ -0,0 +1,265 @@
+    Map<StructLike, V> partitionMap = partitionMaps.computeIfAbsent(specId, this::newPartitionMap);
+    return partitionMap.put(struct, value);
+  }
+
+  @Override
+  public void putAll(Map<? extends Pair<Integer, StructLike>, ? extends V> otherMap) {
+    otherMap.forEach(this::put);
+  }
+
+  @Override
+  public V remove(Object key) {
+    if (key instanceof Pair) {
+      Object first = ((Pair<?, ?>) key).first();
+      Object second = ((Pair<?, ?>) key).second();
+      if (first instanceof Integer && second instanceof StructLike) {
+        return removeKey((Integer) first, (StructLike) second);
+      }
+    }
+
+    return null;
+  }
+
+  public V removeKey(int specId, StructLike struct) {
+    Map<StructLike, V> partitionMap = partitionMaps.get(specId);
+    return partitionMap != null ? partitionMap.remove(struct) : null;
+  }
+
+  @Override
+  public void clear() {
+    partitionMaps.clear();
+  }
+
+  @Override
+  public Set<Pair<Integer, StructLike>> keySet() {
+    PartitionSet keySet = PartitionSet.create(specs);
+
+    for (Entry<Integer, Map<StructLike, V>> specIdAndPartitionMap : partitionMaps.entrySet()) {
+      int specId = specIdAndPartitionMap.getKey();
+      Map<StructLike, V> partitionMap = specIdAndPartitionMap.getValue();
+      for (StructLike partition : partitionMap.keySet()) {
+        keySet.add(specId, partition);
+      }
+    }
+
+    return Collections.unmodifiableSet(keySet);
+  }
+
+  @Override
+  public Collection<V> values() {
+    List<V> values = Lists.newArrayList();
+
+    for (Map<StructLike, V> partitionMap : partitionMaps.values()) {
+      values.addAll(partitionMap.values());
+    }
+
+    return Collections.unmodifiableCollection(values);
+  }
+
+  @Override
+  public Set<Entry<Pair<Integer, StructLike>, V>> entrySet() {
+    Set<Entry<Pair<Integer, StructLike>, V>> entrySet = Sets.newHashSet();
+
+    for (Entry<Integer, Map<StructLike, V>> specIdAndPartitionMap : partitionMaps.entrySet()) {
+      int specId = specIdAndPartitionMap.getKey();
+      Map<StructLike, V> partitionMap = specIdAndPartitionMap.getValue();
+      for (Entry<StructLike, V> structAndValue : partitionMap.entrySet()) {
+        StructLike struct = structAndValue.getKey();
+        V value = structAndValue.getValue();
+        entrySet.add(new PartitionEntry<>(specId, struct, value));
+      }
+    }
+
+    return Collections.unmodifiableSet(entrySet);
+  }
+
+  public V computeIfAbsent(int specId, StructLike struct, Supplier<V> valueSupplier) {
+    Map<StructLike, V> partitionMap = partitionMaps.computeIfAbsent(specId, this::newPartitionMap);
+    return partitionMap.computeIfAbsent(struct, key -> valueSupplier.get());
+  }
+
+  private Map<StructLike, V> newPartitionMap(int specId) {
+    PartitionSpec spec = specs.get(specId);
+    return StructLikeMap.create(spec.partitionType());
+  }
+
+  @Override
+  public String toString() {
+    return partitionMaps.entrySet().stream()
```

Comment:
Why aren't we using this.entryset()?

Reply:
We could but I didn't do that for a few reasons:
- Overhead of constructing `entrySet` as we need to create `PartitionEntry`.
- The need to look up `PartitionSpec` for every pair vs once per partition map right now.